pkg/volume/csi/csi_block.go

```diff
@@ -299,26 +297,24 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
 
 	glog.V(4).Infof(log("unmapper.TearDownDevice(globalMapPath=%s; devicePath=%s)", globalMapPath, devicePath))
 
-	csi := m.csiClient
 	ctx, cancel := context.WithTimeout(context.Background(), csiTimeout)
 	defer cancel()
 
-	// unmap global device map path
-	if err := csi.NodeUnstageVolume(ctx, m.volumeID, globalMapPath); err != nil {
-		glog.Errorf(log("blockMapper.TearDownDevice failed: %v", err))
+	// Call NodeUnpublishVolume
+	publishPath := m.getPublishPath()
+	if _, err := os.Stat(publishPath); err != nil {
+		glog.V(4).Infof(log("blockMapper.TearDownDevice publishPath(%s) has already been deleted, skip calling NodeUnpublishVolume", publishPath))
+	} else if err := m.unpublishVolumeForBlock(ctx, m.csiClient, publishPath); err != nil {
 		return err
 	}
-	glog.V(4).Infof(log("blockMapper.TearDownDevice NodeUnstageVolume successfully [%s]", globalMapPath))
 
-	// request to remove pod volume map path also
-	podVolumePath, volumeName := m.GetPodDeviceMapPath()
-	podVolumeMapPath := filepath.Join(podVolumePath, volumeName)
-	if err := csi.NodeUnpublishVolume(ctx, m.volumeID, podVolumeMapPath); err != nil {
-		glog.Error(log("blockMapper.TearDownDevice failed: %v", err))
+	// Call NodeUnstageVolume
+	stagingPath := m.getStagingPath()
+	if _, err := os.Stat(stagingPath); err != nil {
```

Comment:
Decouple this if-else blocks since they (os.Stat check and unstageVol call) are independent.  In addition the stat check should only fail if err is not IsNotExist(err)

Suggestion:

```go
if _, err := os.Stat(...); os.IsNotExist(err) {
    glog.V(4).Infof(log("blockMapper.TearDownDevice stagingPath(%s) has already been deleted..."))
} else{
    return err
}
if err := m.unstageVolumeForBlock(ctx, m.csiClient, stagingPath); err != nil {
    return err
}
```

Reply:
@vladimirvivien 

Thank you for your suggestion.

I agree that os.stat check should only fail if err is not IsNotExist(err).

For IsNotExist(err) case, I intended to skip calling unstageVolumeForBlock because main purpose for unstageVolumeForBlock is to make staging path invisible, but the path does not exist already.

So, I fixed it as so. And I also applied the same change to unpublishVolumeForBlock.

PTAL